src/commonMain/kotlin/fr/acinq/lightning/channel/InteractiveTx.kt

```diff
@@ -829,44 +837,54 @@ data class InteractiveTxSigningSession(
-                        channelKeys,
-                        remoteCommitmentIndex,
-                        channelParams.localParams,
-                        channelParams.remoteParams,
-                        fundingTxIndex,
-                        fundingParams.remoteFundingPubkey,
-                        firstCommitTx.remoteCommitTx.input,
-                        remotePerCommitmentPoint,
-                        alternativeSpec
-                    )
-                    val alternativeSig = Transactions.sign(alternativeRemoteCommitTx, channelKeys.fundingKey(fundingTxIndex))
-                    CommitSigTlv.AlternativeFeerateSig(feerate, alternativeSig)
+                val localSigOfRemoteCommitTx = Transactions.sign(firstCommitTx.remoteCommitTx, channelKeys.fundingKey(fundingTxIndex))
+                val sortedHtlcTxs = firstCommitTx.remoteHtlcTxs.sortedBy { it.input.outPoint.index }
+                val localSigsOfRemoteHtlcTxs = sortedHtlcTxs.map { Transactions.sign(it, channelKeys.htlcKey.deriveForCommitment(remotePerCommitmentPoint), SigHash.SIGHASH_SINGLE or SigHash.SIGHASH_ANYONECANPAY) }
+
+                val alternativeSigs = if (firstCommitTx.remoteHtlcTxs.isEmpty()) {
+                    val commitSigTlvs = Commitments.alternativeFeerates.map { feerate ->
+                        val alternativeSpec = firstCommitTx.remoteSpec.copy(feerate = feerate)
+                        val (alternativeRemoteCommitTx, _) = Commitments.makeRemoteTxs(
+                            channelKeys,
+                            remoteCommitmentIndex,
+                            channelParams.localParams,
+                            channelParams.remoteParams,
+                            fundingTxIndex,
+                            fundingParams.remoteFundingPubkey,
+                            firstCommitTx.remoteCommitTx.input,
+                            remotePerCommitmentPoint,
+                            alternativeSpec
+                        )
+                        val sig = Transactions.sign(alternativeRemoteCommitTx, channelKeys.fundingKey(fundingTxIndex))
+                        CommitSigTlv.AlternativeFeerateSig(feerate, sig)
+                    }
+                    TlvStream(CommitSigTlv.AlternativeFeerateSigs(commitSigTlvs) as CommitSigTlv)
+                } else {
+                    TlvStream.empty()
                 }
-                val commitSig = CommitSig(channelParams.channelId, localSigOfRemoteTx, listOf(), TlvStream(CommitSigTlv.AlternativeFeerateSigs(alternativeSigs)))
-                val unsignedLocalCommit = UnsignedLocalCommit(localCommitmentIndex, firstCommitTx.localSpec, firstCommitTx.localCommitTx, listOf())
+                val commitSig = CommitSig(channelParams.channelId, localSigOfRemoteCommitTx, localSigsOfRemoteHtlcTxs, alternativeSigs)
+                val unsignedLocalCommit = UnsignedLocalCommit(localCommitmentIndex, firstCommitTx.localSpec, firstCommitTx.localCommitTx, sortedHtlcTxs)
```

Comment:
I don't understand why `firstCommitTx.localHtlcTxs` should not be used here, except to reduce serialization size. This is the list of htlcs that are waiting for the `CommitSig` from our remote peer, but we should have all the information we need to create unsigned htlc txs, Right?

That said, because we recreate those htlc txs in `fromCommitSig` when we receive their `CommitSig` there's no need to set them here.

Which makes me think...  I couldn't find anywhere that `UnsignedLocalCommit.htlcTxs` is used. 

Do we even need the `htlcTxs` parameter at all now in `UnsignedLocalCommit`? I think it can be removed. Though that would require updating serialization and might not be worth the added discriminator logic.

Reply:
> Do we even need the htlcTxs parameter at all now in UnsignedLocalCommit? I think it can be removed. Though that would require updating serialization and might not be worth the added discriminator logic.

Yes, that's exactly why: we added this field early on because we though we'd need it, and wanted to be future-proof to avoid changing the codecs when adding support for splice with htlcs. But it turns out that this is never used and we should have removed that field, but removing would require a codec change, so we instead just always keep it empty.

Filling it with the unsigned local htlc transactions would not be useful, and would only waste space in the backups, that's why we don't bother and always set this to `listOf()`.